logic/src/androidMain/kotlin/com/wire/kalium/logic/network/NetworkStateObserverImpl.kt

```diff
@@ -25,40 +25,75 @@ import android.net.NetworkCapabilities
+    connectivityManager: ConnectivityManager,
+    kaliumDispatcher: KaliumDispatcher = KaliumDispatcherImpl,
+) : NetworkStateObserver {
+
+    constructor(
+        appContext: Context,
+        kaliumDispatcher: KaliumDispatcher = KaliumDispatcherImpl,
+    ) : this(
+        connectivityManager = appContext.getSystemService(Activity.CONNECTIVITY_SERVICE) as ConnectivityManager,
+        kaliumDispatcher = kaliumDispatcher
+    )
+    private val defaultNetworkDataStateFlow: MutableStateFlow<DefaultNetworkData?>
+    private val networkStateFlow: StateFlow<NetworkState>
+    private val scope = CoroutineScope(SupervisorJob() + kaliumDispatcher.default)
 
     init {
-        val initialState = connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork).toState()
-        networkStateFlow = MutableStateFlow(initialState)
+        val defaultNetwork = connectivityManager.activeNetwork
+        val defaultNetworkCapabilities = connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork)
+        defaultNetworkDataStateFlow = MutableStateFlow(defaultNetwork?.let { DefaultNetworkData(it, defaultNetworkCapabilities) })
+        val initialState = defaultNetwork?.let { defaultNetworkCapabilities.toState() } ?: NetworkState.NotConnected
+        networkStateFlow = defaultNetworkDataStateFlow
+            .map { networkData ->
+                when {
+                    networkData == null -> NetworkState.NotConnected
+                    networkData.isBlocked -> NetworkState.ConnectedWithoutInternet
+                    else -> networkData.networkCapabilities.toState()
+                }
+            }
+            .stateIn(scope, SharingStarted.Eagerly, initialState)
 
         val callback = object : ConnectivityManager.NetworkCallback() {
 
             override fun onCapabilitiesChanged(network: Network, networkCapabilities: NetworkCapabilities) {
                 super.onCapabilitiesChanged(network, networkCapabilities)
-                val networkState = networkCapabilities.toState()
-                kaliumLogger.i("${NetworkStateObserver.TAG} capabilities changed $networkState")
-                networkStateFlow.tryEmit(networkState)
+                kaliumLogger.i(
+                    "${NetworkStateObserver.TAG} capabilities changed " +
+                            "internet:${networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)} " +
+                            "validated:${networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)}"
+                )
+                defaultNetworkDataStateFlow.update { DefaultNetworkData(network, networkCapabilities) }
             }
 
             override fun onLost(network: Network) {
                 kaliumLogger.i("${NetworkStateObserver.TAG} lost connection")
-                networkStateFlow.tryEmit(NetworkState.NotConnected)
+                defaultNetworkDataStateFlow.update { null }
```

Comment:
question: why null?

Reply:
`null` is passed when there is no network available at the given moment. I could add another field to the `DefaultNetworkData`, something like `isAvailable` - @mchenani do you think that would be better? I just thought `null` would be more straightforward - if null then no network available, else there is a network connected 😄

Reply:
Thanks for the clarification, IMO I think for  better readability, it's better to avoid having cases as null. Thanks

Reply:
Ok, thanks @mchenani you're right, I replaced `null` with a sealed class